Fixes and improvements to existing sheared components

frenzibyte opened the pull request:
- Removed padding from `ShearedButton` for better sheared flow alignment, as per https://github.com/ppy/osu/commit/1e2af5af4341c073a50feb0e71b324dcd4942ab5. Relevant test case added in `TestSceneShearedButtons`

- Added drop shadow support for `ShearedNub`, to be used in filter control's difficulty range slider (also bundled https://github.com/ppy/osu/pull/32847/commits/715396e5c476c18788169cf336e77396e8be6cc5 to be used for same component). Relevant test case added in `TestSceneShearedSliderBar`.

- Fixed `ShearedSliderBar` left/right boxes not being scaled correctly.

  Before:

  https://github.com/user-attachments/assets/69c4f551-9445-4896-87d2-c163e953874e

frenzibyte committed:
Remove padding from `ShearedButton` for better sheared flow alignment

frenzibyte committed:
Add drop shadow support in `ShearedNub`

Used for range sliders

frenzibyte committed:
Allow disabling focus indicator in `ShearedSliderBar`

frenzibyte committed:
Update sheared button flow test case to be useful

frenzibyte committed:
Update sheared slider bar test scene

frenzibyte committed:
Fix `ShearedSliderBar` left/right boxes not resized correctly

frenzibyte committed:
Add slider step for resizing nub width

frenzibyte committed:
Remove weird default for slider nub X value

peppy commented:
At least one existing usage broke:

![osu! 2025-04-22 at 07 59 47](https://github.com/user-attachments/assets/f4763228-1367-4186-bdfa-3c803393f918)

peppy committed:
Add padding around footer content to avoid sheared overflow

peppy commented:
Not 100% sure about 62b9646, it breaks cases where sheared buttons aren't aligned with other sheared elements:

![osu! 2025-04-22 at 08 23 50](https://github.com/user-attachments/assets/0769eb42-bd98-44bf-b2b7-0592b57226b2)

peppy committed:
Make mod preset shear buttons take up full width

peppy reviewed (commented):
As commented. I fixed one case locally, but I'm not 100% on the direction here.

frenzibyte commented:
> > Before:
> 
> 
> 
> After is missing 😄 

I just wanted to show what the breakage was, "after" should be imaginable.

frenzibyte commented:
> Not 100% sure about [62b9646](https://github.com/ppy/osu/commit/62b96466c4ee8d5c267de6276879656eccc215c0), it breaks cases where sheared buttons aren't aligned with other sheared elements:
> 
> ![osu! 2025-04-22 at 08 23 50](https://private-user-images.githubusercontent.com/191335/435977796-0769eb42-bd98-44bf-b2b7-0592b57226b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU2MjI1MDksIm5iZiI6MTc0NTYyMjIwOSwicGF0aCI6Ii8xOTEzMzUvNDM1OTc3Nzk2LTA3NjllYjQyLWJkOTgtNDRiZi1iMmI3LTA1OTJiNTcyMjZiMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNVQyMzAzMjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZmNhYzIyNmJiMjczY2RmYmM1OWVhZjg4MWQ4NWQ1MzMwY2U5YWY3MWYxYWQzM2Q3NGY3MmE5YWM0ODRjY2QxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.S_jqqpy5l1TuGGLEuphjvUyxR9EwAgUkMxMjokMKZ_M)
> 
> I don't even know how to fix these cases with the current structure.

@peppy How opposed would you be to using non-sheared buttons for cases where buttons need to be aligned vertically. Sheared buttons cannot be aligned vertically without the flow container being sheared and the buttons be aligned in a sheared manner.

With `RoundedButton` for example:

![CleanShot 2025-04-26 at 02 22 22](https://github.com/user-attachments/assets/4af6a09b-0d1f-4410-b1c0-ec3451e5ac64)

peppy commented:
Let's go with the latter diff, I think it reads better than the weird "0" size specification and override stuff 👍

peppy commented:
Next issue: animation is masked, look at the right edge of the button in the video


https://github.com/user-attachments/assets/668e9776-d18f-4be7-a6fe-afba2219e785

peppy reviewed (commented):
Still has issues

frenzibyte committed:
Fix sheared button alignment in preset popovers

frenzibyte committed:
Fix weird alignment code in wizard overlay footer content

frenzibyte committed:
Fix sheared button getting cut when scaled beyond 100%

Keep masking back in `Content`, since the scaling animation is happening
on `Content` instead of `this`. This doesn't regress the intended
behaviour in this PR (which is to just to make the button class itself
sheared instead of its content).

frenzibyte commented:
Applied vertical button alignment fix in https://github.com/ppy/osu/pull/32847/commits/5599a0bc6d2957ab6e71787b23d57b99e2343790, fixed issue noted above in https://github.com/ppy/osu/pull/32847/commits/2924744a73327d85b477339c4cc0343a5c2a5065, and also adjusted alignment code in wizard overlay footer content to make sense in https://github.com/ppy/osu/pull/32847/commits/b22f03a20b95bb8e55357a0cceaf3c963b34af0c.

peppy commented:
```diff 
diff --git a/osu.Game/Overlays/Mods/AddPresetPopover.cs b/osu.Game/Overlays/Mods/AddPresetPopover.cs
index 40a1e4f7e9..31d9ce31fb 100644
--- a/osu.Game/Overlays/Mods/AddPresetPopover.cs
+++ b/osu.Game/Overlays/Mods/AddPresetPopover.cs
@@ -40,11 +40,13 @@ internal partial class AddPresetPopover : OsuPopover
 
         public AddPresetPopover(AddPresetButton addPresetButton)
         {
+            const float content_width = 300;
+
             button = addPresetButton;
 
             Child = new FillFlowContainer
             {
-                Width = 300,
+                Width = content_width,
                 AutoSizeAxes = Axes.Y,
                 Spacing = new Vector2(7),
                 Children = new Drawable[]
@@ -63,12 +65,12 @@ public AddPresetPopover(AddPresetButton addPresetButton)
                         Label = CommonStrings.Description,
                         TabbableContentContainer = this
                     },
-                    createButton = new ShearedButton(0)
+                    createButton = new ShearedButton(content_width)
                     {
-                        Anchor = Anchor.TopCentre,
-                        Origin = Anchor.TopCentre,
+                        // todo: for some very odd reason, this needs to be anchored to topright for the fill flow to be correctly sized to the AABB of the sheared button
+                        Anchor = Anchor.TopRight,
+                        Origin = Anchor.TopRight,
                         RelativeSizeAxes = Axes.X,
-                        Width = 1,
                         Text = ModSelectOverlayStrings.AddPreset,
                         Action = createPreset
                     }
diff --git a/osu.Game/Overlays/Mods/EditPresetPopover.cs b/osu.Game/Overlays/Mods/EditPresetPopover.cs
index 8295bdbab8..eb128c7792 100644
--- a/osu.Game/Overlays/Mods/EditPresetPopover.cs
+++ b/osu.Game/Overlays/Mods/EditPresetPopover.cs
@@ -52,9 +52,11 @@ public EditPresetPopover(Live<ModPreset> preset)
         [BackgroundDependencyLoader]
         private void load()
         {
+            const float content_width = 300;
+
             Child = new FillFlowContainer
             {
-                Width = 300,
+                Width = content_width,
                 AutoSizeAxes = Axes.Y,
                 Spacing = new Vector2(7),
                 Direction = FillDirection.Vertical,
@@ -107,29 +109,27 @@ private void load()
                     {
                         Anchor = Anchor.TopCentre,
                         Origin = Anchor.TopCentre,
-                        RelativeSizeAxes = Axes.X,
-                        AutoSizeAxes = Axes.Y,
+                        AutoSizeAxes = Axes.Both,
                         Spacing = new Vector2(7),
+                        Direction = FillDirection.Vertical,
                         Children = new Drawable[]
                         {
-                            useCurrentModsButton = new ShearedButton(0)
+                            useCurrentModsButton = new ShearedButton(content_width)
                             {
-                                Anchor = Anchor.TopCentre,
-                                Origin = Anchor.TopCentre,
-                                RelativeSizeAxes = Axes.X,
-                                Width = 1,
+                                // todo: for some very odd reason, this needs to be anchored to topright for the fill flow to be correctly sized to the AABB of the sheared button
+                                Anchor = Anchor.TopRight,
+                                Origin = Anchor.TopRight,
                                 Text = ModSelectOverlayStrings.UseCurrentMods,
                                 DarkerColour = colours.Blue1,
                                 LighterColour = colours.Blue0,
                                 TextColour = colourProvider.Background6,
                                 Action = useCurrentMods,
                             },
-                            saveButton = new ShearedButton(0)
+                            saveButton = new ShearedButton(content_width)
                             {
-                                Anchor = Anchor.TopCentre,
-                                Origin = Anchor.TopCentre,
-                                RelativeSizeAxes = Axes.X,
-                                Width = 1,
+                                // todo: for some very odd reason, this needs to be anchored to topright for the fill flow to be correctly sized to the AABB of the sheared button
+                                Anchor = Anchor.TopRight,
+                                Origin = Anchor.TopRight,
                                 Text = Resources.Localisation.Web.CommonStrings.ButtonsSave,
                                 DarkerColour = colours.Orange1,
                                 LighterColour = colours.Orange0,
```
